packages/components/icon/src/Icon.tsx

```diff
@@ -0,0 +1,135 @@
+import { css, cx } from 'emotion';
+import React, { forwardRef } from 'react';
+import tokens from '@contentful/f36-tokens';
+import { getSpacingStyles, Primitive } from '@contentful/f36-core';
+import type {
+  CommonProps,
+  ComponentVariant,
+  MarginProps,
+  PaddingProps,
+  Simplify,
+} from '@contentful/f36-core';
+import type {
+  ComponentType,
+  ExoticComponent,
+  ReactElement,
+  SVGAttributes,
+} from 'react';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type IconComponent = ExoticComponent<any> | ComponentType<any>;
+
+export type IconSize = 'large' | 'medium' | 'small' | 'tiny';
+
+export type IconVariant = Simplify<ComponentVariant | 'muted' | 'white'>;
+
+const sizes: { [key in IconSize]: { [key in 'height' | 'width']: string } } = {
+  large: {
+    height: '32px',
+    width: '32px',
+  },
+  medium: {
+    height: '24px',
+    width: '24px',
+  },
+  small: {
+    height: '18px',
+    width: '18px',
+  },
+  tiny: {
+    height: '16px',
+    width: '16px',
+  },
+};
+
+const fills: { [key in IconVariant]: string } = {
+  muted: tokens.colorTextLight,
+  negative: tokens.colorNegative,
+  positive: tokens.colorPositive,
+  primary: tokens.colorBlueBase,
+  secondary: tokens.colorContrastMid,
+  warning: tokens.colorWarning,
+  white: tokens.colorWhite,
+};
+
+type AsOrChildren =
+  | {
+      as: IconComponent;
+      children?: never;
+    }
+  | {
+      as?: never;
+      children: ReactElement | ReactElement[];
+    };
+
+type IconInternalProps = {
+  /**
+   * Determines the size of the icon
+   */
+  size?: IconSize;
+  /**
+   * Whether or not to trim the icon width, i.e. set `width` to `auto`
+   */
+  trimmed?: boolean;
+  /**
+   * Determines the fill color used
+   */
+  variant?: IconVariant;
+  /**
+   * Custom SVG viewBox attribute to use
+   */
+  viewBox?: SVGAttributes<SVGSVGElement>['viewBox'];
+} & AsOrChildren;
+
+export type IconProps = IconInternalProps &
+  CommonProps &
+  MarginProps &
+  PaddingProps &
+  SVGAttributes<SVGSVGElement>;
+
+export const Icon = forwardRef<SVGSVGElement, IconProps>(function Icon(
```

Comment:
❗ I guess the last major problem to solve is to fix polymorphic typings problem:


```jsx
function MyCustomIcon(props: { foo: string }) {
   return <svg>...</svg>
}

<Icon as={MyCustomIcon} foo="bar" /> 
// this should be valid
// because actually all ...rest props are passed in <MyCustomIcon {...rest} />

```